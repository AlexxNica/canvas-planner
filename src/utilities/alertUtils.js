let visualSuccessFunc = null;
let visualErrorFunc = null;
let srAlertFunc = null;

export function initialize ({ visualSuccessCallback = visualSuccessFunc, visualErrorCallback = visualErrorFunc, srAlertCallback = srAlertFunc }) {
  visualSuccessFunc = visualSuccessCallback;
  visualErrorFunc = visualErrorCallback;
  srAlertFunc = srAlertCallback;
}

export function alert (message, isError = false) {
  if (isError) {
    visualErrorFunc(message);
  } else {
    visualSuccessFunc(message);
  }
}

export function srAlert (message) {
  srAlertFunc(message);
}
