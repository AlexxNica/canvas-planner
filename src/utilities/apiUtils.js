import moment from 'moment-timezone';
import _ from 'lodash';

const getItemDetailsFromPlannable = (apiResponse) => {
  const { plannable, plannable_type, planner_override } = apiResponse;
  const markedComplete = planner_override && planner_override.marked_complete;
  const details = {
    course_id: plannable.course_id,
    title: plannable.name || plannable.title,
    date: plannable.due_at || plannable.todo_date,
    completed: (markedComplete != null) ? markedComplete : plannable.has_submitted_submissions, // TODO: Fix this to use the status field
    points: plannable.points_possible,
    html_url: plannable.html_url,
    overrideId: planner_override && planner_override.id,
    overrideAssignId: plannable.assignment_id,
    id: plannable.id || plannable.page_id
  };
  if (plannable_type === 'discussion_topic') {
    if (plannable.assignment) {
      details.date = plannable.assignment.due_at;
    }
    details.unread_count = plannable.unread_count;
  }
  if (plannable_type === 'announcement' && !details.date) {
    details.date = plannable.posted_at;
  }

  return details;
};

const TYPE_MAPPING = {
  quiz: "Quiz",
  discussion_topic: "Discussion",
  assignment: "Assignment",
  wiki_page: "Page",
  announcement: "Announcement",
  planner_note: "To Do",
};

const getItemType = (plannableType) => {
  return TYPE_MAPPING[plannableType];
};

const getApiItemType = (overrideType) => {
  return _.findKey(TYPE_MAPPING, _.partial(_.isEqual, overrideType));
};

/**
* Translates the API data to the format the planner expects
**/
export function transformApiToInternalItem (apiResponse, courses, timeZone) {
  if (timeZone == null) throw new Error('timezone is required when interpreting api data in transformApiToInternalItem');

  const contextInfo = {};
  if (apiResponse.context_type) {
    const contextId = apiResponse[`${apiResponse.context_type.toLowerCase()}_id`];
    const course = courses.find(c => c.id === contextId);
    if (course) {
      contextInfo.context = {
        type: apiResponse.context_type,
        id: contextId,
        title: course.shortName,
        image_url: course.image,
        color: course.color,
        url: course.href
      };
    }
  }

  const details = getItemDetailsFromPlannable(apiResponse);

  if ((!contextInfo.context) && apiResponse.plannable_type === 'planner_note' && (details.course_id)) {
    const course = courses.find(c => c.id === details.course_id);
    contextInfo.context = {
      type: 'Planner Note',
      id: details.course_id,
      title: course.shortName,
      image_url: course.image,
      color: course.color,
      url: course.href
    };
  }

  if (details.unread_count && apiResponse.submissions) {
    apiResponse.submissions.unread_count = details.unread_count;
  }
  return {
    ...contextInfo,
    id: apiResponse.plannable_id,
    dateBucketMoment: moment.tz(details.date, timeZone).startOf('day'),
    type: getItemType(apiResponse.plannable_type),
    status: apiResponse.submissions,
    ...details
  };
}

/**
 * This takes the response from creating a new planner note aka To Do and puts it in the internal
 * format.
 */
export function transformPlannerNoteApiToInternalItem (plannerItemApiResponse, courses, timeZone) {
  const plannerNote = plannerItemApiResponse.planner_note;
  return {
    id: plannerNote.id,
    dateBucketMoment: moment.tz(plannerNote.todo_date, timeZone).startOf('day'),
    type: 'To Do',
    status: false,
    course_id: plannerNote.course_id,
    context: {
      id: plannerNote.course_id
    },
    title: plannerNote.title,
    date: plannerNote.todo_date,
    details: plannerNote.details,
    completed: false
  };
}

/**
* Turn internal item format into data the API can consume for save actions
*/
export function transformInternalToApiItem (internalItem) {
  const contextInfo = {};
  if (internalItem.context) {
    contextInfo.context_type = internalItem.context.type || 'Course';
    contextInfo[`${contextInfo.context_type.toLowerCase()}_id`] = internalItem.context.id;
  }
  return {
    id: internalItem.id,
    ...contextInfo,
    todo_date: internalItem.date,
    title: internalItem.title,
    details: internalItem.details,
  };
}

export function transformInternalToApiOverride (internalItem, userId) {
  let type = getApiItemType(internalItem.type);
  let id = internalItem.id;
  if (internalItem.overrideAssignId) {
    type = 'assignment';
    id = internalItem.overrideAssignId;
  }
  return {
    id: internalItem.overrideId,
    plannable_id: id,
    plannable_type: type,
    user_id: userId,
    marked_complete: internalItem.completed
  };
}

