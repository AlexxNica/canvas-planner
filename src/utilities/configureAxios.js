export default function (axiosInstance) {
  // Add CSRF stuffs to make Canvas happy when we are making requests with axios
  axiosInstance.defaults.xsrfCookieName = '_csrf_token';
  axiosInstance.defaults.xsrfHeaderName = 'X-CSRF-Token';

  // Handle stringified IDs for JSON responses
  var originalDefaults = axiosInstance.defaults.headers.common['Accept'];
  axiosInstance.defaults.headers.common['Accept'] = 'application/json+canvas-string-ids, ' + originalDefaults;

  // Rails checks this header to decide if a request is an xhr request
  axiosInstance.defaults.headers.common['X-Requested-With'] = 'XMLHttpRequest';

  return axiosInstance;
}
