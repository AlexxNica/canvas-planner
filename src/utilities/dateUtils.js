import moment from 'moment-timezone';
import formatMessage from '../format-message';

function getTodaysDetails () {
  const today = moment();
  const yesterday = today.clone().subtract(1, 'days');
  const tomorrow = today.clone().add(1, 'days');

  return { today, yesterday, tomorrow };
}

function isSpecialDay (date) {
  const { today, yesterday, tomorrow } = getTodaysDetails();
  const momentizedDate = new moment(date);

  const specialDates = [today, yesterday, tomorrow];
  return specialDates.some(sd => sd.isSame(momentizedDate, 'day'));
}

export function isToday (date, today = moment()) {
  const momentizedDate = new moment(date);
  return today.isSame(momentizedDate, 'day');
}

export function isInPast (date, today = moment()) {
  const momentizedDate = new moment(date);
  return momentizedDate.isBefore(today, 'day');
}

/**
* Given a date (in any format that moment will digest)
* it will return a string indicating Today, Tomorrow, Yesterday
* or the day of the week if it doesn't fit in any of those categories
*/
export function getFriendlyDate (date) {
  const { today, yesterday, tomorrow } = getTodaysDetails();
  const momentizedDate = new moment(date);

  if (isToday(date, today)) {
    return formatMessage('Today');
  } else if (yesterday.isSame(momentizedDate, 'day')) {
    return formatMessage('Yesterday');
  } else if (tomorrow.isSame(momentizedDate, 'day')) {
    return formatMessage('Tomorrow');
  } else {
    return momentizedDate.format('dddd');
  }
}


export function getFullDate (date) {
  if (isSpecialDay(date)) {
    return moment(date).format('dddd, MMMM D');
  } else {
    return moment(date).format('MMMM D, YYYY');
  }
}

export function formatDayKey (date) {
  return moment(date, moment.ISO_8601).format('YYYY-MM-DD');
}

export function getFirstLoadedMoment (days, timeZone) {
  if (!days.length) return moment().tz(timeZone).startOf('day');
  const firstLoadedDay = days[0];
  const firstLoadedItem = firstLoadedDay[1][0];
  return firstLoadedItem.dateBucketMoment.clone();
}

export function getLastLoadedMoment (days, timeZone) {
  if (!days.length) return moment().tz(timeZone).startOf('day');
  const lastLoadedDay = days[days.length-1];
  const loadedItem = lastLoadedDay[1][0];
  return loadedItem.dateBucketMoment.clone();
}
