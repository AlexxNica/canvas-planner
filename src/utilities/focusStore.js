/**
 * Handles focusing elements when needed between different places
 */
/*
 * Copyright (C) 2017 - present Instructure, Inc.
 *
 * This module is part of Canvas.
 *
 * This module and Canvas are free software: you can redistribute them and/or modify them under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * This module and Canvas are distributed in the hope that they will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */
export class FocusStore {
  itemToFocus = null;

  setItemToFocus (item) {
    if (typeof item.focus !== "function") {
      throw new Error('Focus Store: You can only set items to focus that have a focus method.');
    }
    this.itemToFocus = item;
  }

  focus () {
    if (this.itemToFocus == null) {
      throw new Error('Focus Store: You need to set an item before it can be focused.');
    }

    // TODO: Make less hacky this: currently is a bandaid over the scrolling when saving a todo issue
    var save = document.body.scrollTop;
    this.itemToFocus.focus();
    document.body.scrollTop = save;

    this.itemToFocus = null;
  }
}

const focusStore = new FocusStore();
export default focusStore;
