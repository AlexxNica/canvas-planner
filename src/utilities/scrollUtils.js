import _ from 'lodash';
import Velocity from 'velocity-animate';

const animationQueue = [];

function runAnimationQueue () {
  while (animationQueue.length) {
    const animationFn = animationQueue.shift();
    animationFn();
  }
}

// Based on this formula:
// element's position in the viewport + the window's scroll position === the element's position in the document
// so if we want the scroll position that will maintain the element in it's current viewport position,
// window scroll position = element's current document position - element's initial viewport position
export function maintainViewportPosition (elt, mocks={document, window}) {
  const elementsInitialPositionInViewport = elt.getBoundingClientRect().top;
  mocks.window.requestAnimationFrame(() => {
    const elementsNewPositionInViewport = elt.getBoundingClientRect().top;
    const documentPositionInViewport = mocks.document.documentElement.getBoundingClientRect().top;
    const elementPositionInDocument = elementsNewPositionInViewport - documentPositionInViewport;
    const newWindowScrollPosition = elementPositionInDocument - elementsInitialPositionInViewport;
    // always want this insta-scroll to happen first so later scroll animations will start in the right place.
    animationQueue.unshift(() => {
      mocks.window.scroll(0, newWindowScrollPosition);
    });
    mocks.window.requestAnimationFrame(runAnimationQueue);
  });
}

export function animateSlideDown (elt) {
  Velocity(elt, 'slideDown');
}

export function animateScroll (elt, offset) {
  animationQueue.push(() => {
    Velocity(elt, 'scroll', {offset: -offset});
  });
  window.requestAnimationFrame(runAnimationQueue);
}

function handleScrollUpAttempt (cb, e) {
  e.preventDefault();
  cb();
}

function handleWindowWheel (cb, wind, e) {
  if (wind.pageYOffset === 0 && e.deltaY < 0) {
    handleScrollUpAttempt(cb, e);
  }
}

function handleWindowScrollKey (cb, wind, e) {
  if (wind.pageYOffset === 0 &&
      (e.key === 'PageUp' || e.key === 'ArrowUp' || e.key === 'Up')) {
    handleScrollUpAttempt(cb, e);
  }
}

export function registerScrollEvents (scrollIntoPastCb, wind = window) {
  // do debounce here so tests aren't dependant on global state
  const boundWindowWheel = _.debounce(
    handleWindowWheel.bind(undefined, scrollIntoPastCb, wind),
  500, {leading: true, trailing: false});
  wind.addEventListener('wheel', boundWindowWheel);

  const boundScrollKey = handleWindowScrollKey.bind(undefined, scrollIntoPastCb, wind);
  wind.addEventListener('keydown', boundScrollKey);
}
