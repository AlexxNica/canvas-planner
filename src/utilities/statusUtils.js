import formatMessage from '../format-message';
import _ from 'lodash';

const PILL_MAPPING = {
  'missing': { id: 'missing', text: formatMessage('Missing'), variant: 'danger' },
  'late': { id: 'late', text: formatMessage('Late'), variant: 'danger' },
  'graded': { id: 'graded', text: formatMessage('Graded') },
  'excused': { id: 'excused', text: formatMessage('Excused') },
  'submitted': { id: 'submitted', text: formatMessage('Submitted') },
  'new_grades': { id: 'new_grades', text: formatMessage('New Grades') },
  'has_feedback': { id: 'new_feedback', text: formatMessage('New Feedback') },
  'new_replies': { id: 'new_replies', text: formatMessage('New Replies') },
};

const newActivityKeys = {
  missing: true,
  new_grades: true,
  has_feedback: true,
  new_replies: true,
};

export function isNewActivityItem (item) {
  return _.entries(item.status).some(([status, value]) => value && newActivityKeys[status]);
}

export function anyNewActivity (items) {
  return items && items.some((item => isNewActivityItem(item)));
}

/**
* Returns an array of pill objects that the particular item
* qualifies to have
*/
export function getBadgesForItem (item) {
  let badges = [];
  if (item.status) {
    badges = Object.keys(item.status).filter(key => item.status[key] && key !== 'needs_grading').map(a => PILL_MAPPING[a]);
  }
  return badges;
}

/**
* Returns an array of pill objects that the items qualify to have
*/
export function getBadgesForItems (items) {
  const badges = [];
  if (items.some(i => i.status && i.status.graded)) {
    badges.push(PILL_MAPPING.new_grades);
  }
  if (items.some(i => i.status && i.status.has_feedback)) {
    badges.push(PILL_MAPPING.has_feedback);
  }
  if (items.some(i => i.status && i.status.unread_count)) {
    badges.push(PILL_MAPPING.new_replies);
  }
  return badges;
}
