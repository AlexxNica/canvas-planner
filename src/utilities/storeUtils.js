/*
 * Copyright (C) 2017 - present Instructure, Inc.
 *
 * This module is part of Canvas.
 *
 * This module and Canvas are free software: you can redistribute them and/or modify them under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3 of the License.
 *
 * This module and Canvas are distributed in the hope that they will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

export function findPlannerItemById (state, id) {
  // payload is the item id. find it
  let plannerItem = null;
  state.some(d => d[1].some(i => {
    if (i.id === id) {
      plannerItem = i;
      return true;
    }
    return false;
  }));
  return plannerItem;
}
