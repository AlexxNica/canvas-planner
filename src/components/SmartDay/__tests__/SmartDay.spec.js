import React from 'react';
import {shallow} from 'enzyme';
import {SmartDay, getLastUnseenDayKeyWithNewActivity} from '../index';
import moment from 'moment-timezone';
import {formatDayKey} from '../../../utilities/dateUtils';

describe('getLastUnseenDayKeyWithNewActivity', () => {
  it ('finds a new activity item', () => {
    const result = getLastUnseenDayKeyWithNewActivity({
      days: [
        ['2017-05-01', [
          {status: {}},
          {status: {new_grades: true}},
        ]],
      ],
    });
    expect(result).toBe('2017-05-01');
  });

  it ('returns nothing if no new activity item', () => {
    const result = getLastUnseenDayKeyWithNewActivity({
      days: [
        ['2017-05-01', [
          {status: {}},
          {status: {}},
        ]],
      ],
    });
    expect(result).toBeNull();
  });

  it ('does not find a new activity item on today', () => {
    const todayKey = formatDayKey(moment());
    const result = getLastUnseenDayKeyWithNewActivity({
      days: [
        [todayKey, [
          {status: {}},
          {status: {new_grades: true}},
        ]],
      ],
    });
    expect(result).toBeNull();
  });

  it ('does not find a new activity item in the future', () => {
    const tomorrowKey = formatDayKey(moment().add(1, 'days'));
    const result = getLastUnseenDayKeyWithNewActivity({
      days: [
        [tomorrowKey, [
          {status: {}},
          {status: {new_grades: true}},
        ]],
      ],
    });
    expect(result).toBeNull();
  });
});

describe('SmartDay', () => {
  it('forwards properties to Day', () => {
    const wrapper = shallow(<SmartDay day="2017-06-19" timeZone="UTC" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('calls animate scroll on mounting if the day key matches', () => {
    const scrollMock = jest.fn();
    const wrapper = shallow(<SmartDay
      day="2017-06-19"
      timeZone="UTC"
      animateScroll={scrollMock}
      newActivityTargetDayKey="2017-06-19"
      stickyOffset={42}
      seekingNewActivity />);
    // enzyme doesn't handle functional refs even with mount, so we have to do this ourselves
    wrapper.instance().rootElement = 'some element';
    wrapper.instance().componentDidMount();
    expect(scrollMock).toHaveBeenCalledWith('some element', 42);
  });

  it('requires seekingNewActivity to scroll', () => {
    const scrollMock = jest.fn();
    const wrapper = shallow(<SmartDay
      day="2017-06-19"
      timeZone="UTC"
      animateScroll={scrollMock}
      newActivityTargetDayKey="2017-06-19"
      stickyOffset={42}
      seekingNewActivity={false} />);
    // enzyme doesn't handle functional refs even with mount, so we have to do this ourselves
    wrapper.instance().rootElement = 'some element';
    wrapper.instance().componentDidMount();
    expect(scrollMock).not.toHaveBeenCalled();
  });

  it('requires day key to match target to scroll', () => {
    const scrollMock = jest.fn();
    const wrapper = shallow(<SmartDay
      day="2017-06-19"
      timeZone="UTC"
      animateScroll={scrollMock}
      newActivityTargetDayKey="2017-06-10"
      stickyOffset={42}
      seekingNewActivity />);
    // enzyme doesn't handle functional refs even with mount, so we have to do this ourselves
    wrapper.instance().rootElement = 'some element';
    wrapper.instance().componentDidMount();
    expect(scrollMock).not.toHaveBeenCalled();
  });


});
