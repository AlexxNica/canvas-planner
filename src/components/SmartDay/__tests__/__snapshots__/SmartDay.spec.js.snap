// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SmartDay forwards properties to Day 1`] = `
<Day
  animateScroll={[Function]}
  day="2017-06-19"
  rootElementRef={[Function]}
  stickyOffset={0}
  timeZone="UTC"
/>
`;
